### ISSUE TITLE
argument error `Inspect.Ecto.Query.expr_to_string/4`

### ISSUE THREAD

**Issue by atomkirk** [elixir-ecto/ecto#2467]
### Environment

* Elixir version (elixir -v): 1.6.2
* Database and version (PostgreSQL 9.4, MongoDB 3.2, etc.): Mysql 5.7.18
* Ecto version (mix deps): 2.1.6
* Database adapter and version (mix deps): mariaex 0.8.3
* Operating system: Mac OS Sierra 

### Current behavior

Here's my query:

```
  def revenue(user) do
    from(
      months in subquery(
        from(
          entry in "time_entries",
          cross_join: num in fragment("(SELECT @num := -1)"),
          select: %{
            month: fragment("DATE_FORMAT(NOW() - INTERVAL (@num := @num + 1) MONTH, '%Y-%m')")
          },
          limit: 13
        )
      ),
      left_join: rev in subquery(
        from(
          [line, je] in lines(user),
          join: ca in "chart_accounts", on: line.chart_account_id == ca.id,
          select: %{
            month: fragment("DATE_FORMAT(?, '%Y-%m')", je.date),
            total: fragment("SUM(IF(? = 0, ? * -1, ?))", line.kind, line.amount, line.amount),
          },
          where: ca.account_id == ^user.account_id and ca.classification in ^ZB.ChartClassifications.sales(),
          group_by: fragment("month")
        )
      ), on: rev.month == months.month,
      select: %{
        month: months.month,
        total: rev.total,
      }
    )
    |> ZB.Repo.all()
  end

  defp lines(user) do
    from(
      line in {"journal_entry_lines", ZB.JournalEntryLine},
      join: je in "journal_entries", on: line.journal_entry_id == je.id,
      where: line.account_id == ^user.account_id
        and je.is_confirmed
        and is_nil(je.deleted_at)
        and is_nil(line.deleted_at)
        and je.currency_code == ^user.account.default_currency_code
    )
  end
```
If `select` is this:

```
      select: %{
        month: months.month,
      }
```
It works. But if I try to reference `total` from the left join, like this:
```
      select: %{
        month: months.month,
        total: rev.total,
      }
```

I get: 

```
08:14:41 phoenix-api.1      |  ** (exit) an exception was raised:
08:14:41 phoenix-api.1      |      ** (ArgumentError) argument error
08:14:41 phoenix-api.1      |          (ecto) lib/ecto/query/inspect.ex:160: Inspect.Ecto.Query.expr_to_string/4
08:14:41 phoenix-api.1      |          (elixir) lib/macro.ex:813: Macro.unary_call/2
08:14:41 phoenix-api.1      |          (elixir) lib/macro.ex:678: Macro.to_string/2
08:14:41 phoenix-api.1      |          (elixir) lib/macro.ex:877: Macro.call_to_string/2
08:14:41 phoenix-api.1      |          (elixir) lib/macro.ex:885: Macro.call_to_string_with_args/3
08:14:41 phoenix-api.1      |          (elixir) lib/macro.ex:686: Macro.to_string/2
08:14:41 phoenix-api.1      |          (elixir) lib/macro.ex:950: anonymous fn/2 in Macro.kw_list_to_string/2
08:14:41 phoenix-api.1      |          (elixir) lib/enum.ex:1365: anonymous fn/4 in Enum.map_join/3
08:14:41 phoenix-api.1      |          (elixir) lib/enum.ex:1899: Enum."-map_join/3-lists^foldl/2-0-"/3
08:14:41 phoenix-api.1      |          (elixir) lib/enum.ex:1899: Enum.map_join/3
08:14:41 phoenix-api.1      |          (elixir) lib/macro.ex:586: Macro.to_string/2
08:14:41 phoenix-api.1      |          (ecto) lib/ecto/query/inspect.ex:133: Inspect.Ecto.Query.kw_expr/3
08:14:41 phoenix-api.1      |          (ecto) lib/ecto/query/inspect.ex:64: Inspect.Ecto.Query.to_list/1
08:14:41 phoenix-api.1      |          (ecto) lib/ecto/query/inspect.ex:34: Inspect.Ecto.Query.to_string/1
08:14:41 phoenix-api.1      |          (ecto) lib/ecto/exceptions.ex:34: Ecto.QueryError.exception/1
08:14:41 phoenix-api.1      |          (elixir) lib/enum.ex:1047: anonymous fn/3 in Enum.flat_map_reduce/3
```

### Expected behavior

If I get rid of the `lines` function and merge the query components from `lines` into the query above, it works. Unfortunately that creates duplicate logic I'd like to avoid and I don't see any reason why this shouldn't work… do you?

**Comment by josevalim** [comment]
It is definitely a bug, since it was not supposed to fail like that, but you need to provide us a smaller sample or an application that allows us to reproduce it. The code that you have still have a lot of dependencies on user fields and things like `^ZB.ChartClassifications.sales()`. If you can replace those by values and provide a failing query it would help a lot.

Also note that you are not on the latest Ecto. This may have been fixed in the latest 2.2.

**Comment by atomkirk** [comment]
I'll close for now since I think it would be a significant of work to make an app that duplicates it and it sounds like you haven't seen it before, so its probably rare
